File: ios/native_workmanager/Sources/native_workmanager/utils/RequestSigner.swift
import CryptoKit
import Foundation

/// HMAC-SHA256 request signer for HTTP workers.
///
/// Signs outgoing `URLRequest` instances in-place with an HMAC-SHA256 digest
/// computed over a canonical message:
///
/// ```
/// METHOD\n
/// URL\n
/// BODY\n          ← only when signBody=true and the request has a body
/// TIMESTAMP       ← only when includeTimestamp=true (Unix ms)
/// ```
///
/// The signature is added as `headerName` (default `X-Signature`), and the
/// timestamp (if enabled) as `X-Timestamp`.
///
/// Usage:
/// ```swift
/// RequestSigner.sign(
///     request: &request,
///     config: signingConfig
/// )
/// ```
@available(iOS 13.0, *)
enum RequestSigner {

    struct Config {
        let secretKey: String
        let headerName: String
        let signaturePrefix: String
        let includeTimestamp: Bool
        let signBody: Bool

        init(
            secretKey: String,
            headerName: String       = "X-Signature",
            signaturePrefix: String  = "",
            includeTimestamp: Bool   = true,
            signBody: Bool           = true
        ) {
            self.secretKey        = secretKey
            self.headerName       = headerName
            self.signaturePrefix  = signaturePrefix
            self.includeTimestamp = includeTimestamp
            self.signBody         = signBody
        }

        /// Parse from a JSON dict (from Dart's `RequestSigning.toMap()`).
        static func from(_ dict: [String: Any]?) -> Config? {
            guard let dict = dict,
                  let key = dict["secretKey"] as? String, !key.isEmpty else {
                return nil
            }
            return Config(
                secretKey:        key,
                headerName:       dict["headerName"]      as? String ?? "X-Signature",
                signaturePrefix:  dict["signaturePrefix"] as? String ?? "",
                includeTimestamp: dict["includeTimestamp"] as? Bool  ?? true,
                signBody:         dict["signBody"]         as? Bool  ?? true
            )
        }
    }

    // MARK: - Public API

    /// Mutates [request] by adding the HMAC-SHA256 signature header (and optional timestamp).
    static func sign(request: inout URLRequest, config: Config) {
        let timestamp = config.includeTimestamp
            ? String(Int64(Date().timeIntervalSince1970 * 1000))
            : nil

        let bodyString: String? = config.signBody
            ? (request.httpBody.flatMap { String(data: $0, encoding: .utf8) })
            : nil

        let message = canonicalMessage(
            method:    request.httpMethod ?? "GET",
            url:       request.url?.absoluteString ?? "",
            body:      bodyString,
            timestamp: timestamp
        )

        guard let signature = hmacSHA256(message: message, key: config.secretKey) else {
            print("RequestSigner: HMAC computation failed — request not signed")
            return
        }

        request.setValue("\(config.signaturePrefix)\(signature)", forHTTPHeaderField: config.headerName)
        if let ts = timestamp {
            request.setValue(ts, forHTTPHeaderField: "X-Timestamp")
        }

        print("RequestSigner: Signed — header=\(config.headerName) ts=\(timestamp ?? "none")")
    }

    // MARK: - Private helpers

    private static func canonicalMessage(
        method: String,
        url: String,
        body: String?,
        timestamp: String?
    ) -> String {
        var msg = "\(method.uppercased())\n\(url)\n"
        if let body = body, !body.isEmpty { msg += "\(body)\n" }
        if let ts = timestamp { msg += ts }
        return msg
    }

    private static func hmacSHA256(message: String, key: String) -> String? {
        let keyData = Data(key.utf8)
        let msgData = Data(message.utf8)
        let hmac = HMAC<SHA256>.authenticationCode(
            for: msgData,
            using: SymmetricKey(data: keyData)
        )
        return Data(hmac).map { String(format: "%02x", $0) }.joined()
    }
}

File: android/src/main/kotlin/dev/brewkits/native_workmanager/workers/utils/RequestSigner.kt
package dev.brewkits.native_workmanager.workers.utils

import android.util.Log
import okhttp3.Request
import okio.Buffer
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * HMAC-SHA256 request signer for HTTP workers.
 *
 * Signs outgoing OkHttp [Request] instances with an HMAC-SHA256 digest
 * computed over a canonical message:
 *
 * ```
 * METHOD\n
 * URL\n
 * BODY\n          ← only when signBody=true and the request has a body
 * TIMESTAMP       ← only when includeTimestamp=true (Unix ms)
 * ```
 *
 * The signature is added as [headerName] (default `X-Signature`), and the
 * timestamp (if enabled) as `X-Timestamp`.
 *
 * Usage:
 * ```kotlin
 * val signed = RequestSigner.sign(
 *     request = request,
 *     secretKey = config.requestSigning.secretKey,
 *     headerName = config.requestSigning.headerName,
 *     signaturePrefix = config.requestSigning.signaturePrefix,
 *     includeTimestamp = config.requestSigning.includeTimestamp,
 *     signBody = config.requestSigning.signBody,
 * )
 * ```
 */
object RequestSigner {

    private const val TAG = "RequestSigner"
    private const val HMAC_ALGORITHM = "HmacSHA256"

    data class Config(
        val secretKey: String,
        val headerName: String = "X-Signature",
        val signaturePrefix: String = "",
        val includeTimestamp: Boolean = true,
        val signBody: Boolean = true,
    )

    /**
     * Sign [request] and return a new [Request] with signature headers added.
     *
     * The original request is not modified.
     */
    fun sign(request: Request, config: Config): Request {
        val timestamp = if (config.includeTimestamp) System.currentTimeMillis().toString() else null

        // Read the request body without consuming it (OkHttp bodies can only be read once).
        val bodyString: String? = if (config.signBody && request.body != null) {
            try {
                val buf = Buffer()
                request.body!!.writeTo(buf)
                buf.readUtf8()
            } catch (e: Exception) {
                Log.w(TAG, "Could not read request body for signing: ${e.message}")
                null
            }
        } else null

        val message = buildCanonicalMessage(
            method    = request.method,
            url       = request.url.toString(),
            body      = bodyString,
            timestamp = timestamp,
        )

        val signature = try {
            hmacSha256(message, config.secretKey)
        } catch (e: Exception) {
            Log.e(TAG, "HMAC signing failed: ${e.message}", e)
            return request  // Return unsigned request rather than crashing
        }

        val builder = request.newBuilder()
            .header(config.headerName, "${config.signaturePrefix}$signature")

        if (timestamp != null) builder.header("X-Timestamp", timestamp)

        Log.d(TAG, "Request signed — header=${config.headerName}, ts=$timestamp")
        return builder.build()
    }

    // ── Helpers ──────────────────────────────────────────────────────────────

    private fun buildCanonicalMessage(
        method: String,
        url: String,
        body: String?,
        timestamp: String?,
    ): String = buildString {
        append(method.uppercase()); append("\n")
        append(url);                append("\n")
        if (!body.isNullOrEmpty()) { append(body); append("\n") }
        if (timestamp != null)      append(timestamp)
    }

    private fun hmacSha256(message: String, key: String): String {
        val mac = Mac.getInstance(HMAC_ALGORITHM)
        mac.init(SecretKeySpec(key.toByteArray(Charsets.UTF_8), HMAC_ALGORITHM))
        val digest = mac.doFinal(message.toByteArray(Charsets.UTF_8))
        return digest.joinToString("") { "%02x".format(it) }
    }

    /**
     * Parse a [RequestSigning] config from a JSON map.
     *
     * Returns null if the map is missing or [secretKey] is blank.
     */
    fun fromMap(map: org.json.JSONObject?): Config? {
        if (map == null) return null
        val key = map.optString("secretKey", "")
        if (key.isBlank()) return null
        return Config(
            secretKey        = key,
            headerName       = map.optString("headerName",       "X-Signature"),
            signaturePrefix  = map.optString("signaturePrefix",  ""),
            includeTimestamp = map.optBoolean("includeTimestamp", true),
            signBody         = map.optBoolean("signBody",         true),
        )
    }
}
